src/azure-cli/azure/cli/command_modules/network/tunnel.py

```diff
@@ -0,0 +1,206 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+# pylint: disable=import-error,unused-import
+
+#
+# Modified from tunnel.py in appservice module
+#
+
+import sys
+import ssl
+import json
+import socket
+import time
+import traceback
+import logging as logs
+from contextlib import closing
+from datetime import datetime
+from threading import Thread
+
+import websocket
+from websocket import create_connection, WebSocket
+
+from msrestazure.azure_exceptions import CloudError
+
+from azure.cli.core._profile import Profile
+from azure.cli.core.util import should_disable_connection_verify
+
+import requests
+import urllib3
+
+from knack.util import CLIError
+from knack.log import get_logger
+logger = get_logger(__name__)
+
+
+# pylint: disable=no-member,too-many-instance-attributes,bare-except,no-self-use
+class TunnelServer:
+    def __init__(self, cli_ctx, local_addr, local_port, bastion, remote_host, remote_port):
+        self.local_addr = local_addr
+        self.local_port = int(local_port)
+        if self.local_port != 0 and not self.is_port_open():
+            raise CLIError('Defined port is currently unavailable')
+        self.bastion = bastion
+        self.remote_host = remote_host
+        self.remote_port = remote_port
+        self.client = None
+        self.ws = None
+        self.last_token = None
+        self.node_id = None
+        self.cli_ctx = cli_ctx
+        logger.info('Creating a socket on port: %s', self.local_port)
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        logger.info('Setting socket options')
+        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        logger.info('Binding to socket on local address and port')
+        self.sock.bind((self.local_addr, self.local_port))
+        if self.local_port == 0:
+            self.local_port = self.sock.getsockname()[1]
+            logger.info('Auto-selecting port: %s', self.local_port)
+        logger.info('Finished initialization')
+
+    def is_port_open(self):
+        is_port_open = False
+        with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
+            if sock.connect_ex(('', self.local_port)) == 0:
+                logger.info('Port %s is NOT open', self.local_port)
+            else:
+                logger.info('Port %s is open', self.local_port)
+                is_port_open = True
+            return is_port_open
+
+    def _get_auth_token(self):
+        profile = Profile(cli_ctx=self.cli_ctx)
+        user = profile.get_current_account_user()
+        # Generate an Azure token with the VSTS resource app id
+        auth_token = profile.get_access_token_for_resource(user, None, None)
```

Comment:
`get_access_token_for_resource` will be deprecated pretty soon (https://github.com/Azure/azure-cli/issues/18944). Please migrate to `get_raw_token`:

https://github.com/Azure/azure-cli/blob/14cc787d0f58bc649d402b486fdecc5625eee9ac/src/azure-cli-core/azure/cli/core/_profile.py#L679

Reply:
Why is the third argument `resource` set as `None`?

https://github.com/Azure/azure-cli/blob/14cc787d0f58bc649d402b486fdecc5625eee9ac/src/azure-cli-core/azure/cli/core/_profile.py#L531-L535

AAD will default it to `00000002-0000-0000-c000-000000000000` which is AD Graph API.

Reply:
The only other place where `get_access_token_for_resource` is called is `vsts_cd_provider.py`:

https://github.com/Azure/azure-cli/blob/0a22ec50affe656cd95a210890beb860425523ee/src/azure-cli/azure/cli/command_modules/appservice/vsts_cd_provider.py#L33

`vsts_cd_provider.py` has been removed by #18203:

```
> git log -- src/azure-cli/azure/cli/command_modules/appservice/vsts_cd_provider.py
commit 8701d602677a6e26f3ab681c2ee590879ff2d5c7
Author: Calvin <calvinch4n@gmail.com>
Date:   Thu Jul 1 18:10:37 2021 -0700

    [AppService] BREAKING CHANGE: az webapp deployment source config: Remove vsts-cd-manager (#18203)

    * Remove vsts-cd-manager

    * Rerun tests

    * Remove vsts-cd-manager

    * Rerun tests
```